core/resp.go

```diff
@@ -2,134 +2,124 @@ package core
 
 import (
 	"bytes"
-	"errors"
 	"fmt"
+	"io"
+	"strconv"
 )
 
-// reads the length typically the first integer of the string
-// until hit by an non-digit byte and returns
-// the integer and the delta = length + 2 (CRLF)
-// TODO: Make it simpler and read until we get `\r` just like other functions
-func readLength(data []byte) (int, int) {
-	pos, length := 0, 0
-	for pos = range data {
-		b := data[pos]
-		if !(b >= '0' && b <= '9') {
-			return length, pos + 2
-		}
-		length = length*10 + int(b-'0')
+func readLength(buf *bytes.Buffer) (int64, error) {
+	s, err := readStringUntilSr(buf)
+	if err != nil {
+		return 0, err
 	}
-	return 0, 0
-}
 
-// reads a RESP encoded simple string from data and returns
-// the string, the delta, and the error
-func readSimpleString(data []byte) (string, int, error) {
-	// first character +
-	pos := 1
+	v, err := strconv.ParseInt(s, 10, 64)
+	if err != nil {
+		return 0, err
+	}
 
-	for ; data[pos] != '\r'; pos++ {
+	return v, nil
+}
+
+func readStringUntilSr(buf *bytes.Buffer) (string, error) {
+	s, err := buf.ReadString('\r')
+	if err != nil {
+		return "", err
 	}
+	// increamenting to skip `\n`
+	buf.ReadByte()
+	return s[:len(s)-1], nil
+}
 
-	return string(data[1:pos]), pos + 2, nil
+// reads a RESP encoded simple string from data and returns
+// the string and the error
+// the function internally manipulates the buffer pointer
+// and keepts it at a point where the subsequent value can be read from.
+func readSimpleString(c io.ReadWriter, buf *bytes.Buffer) (string, error) {
+	return readStringUntilSr(buf)
 }
 
 // reads a RESP encoded error from data and returns
-// the error string, the delta, and the error
-func readError(data []byte) (string, int, error) {
-	return readSimpleString(data)
+// the error string and the error
+// the function internally manipulates the buffer pointer
+// and keepts it at a point where the subsequent value can be read from.
+func readError(c io.ReadWriter, buf *bytes.Buffer) (string, error) {
+	return readStringUntilSr(buf)
 }
 
 // reads a RESP encoded integer from data and returns
-// the intger value, the delta, and the error
-func readInt64(data []byte) (int64, int, error) {
-	// first character :
-	pos := 1
-	var value int64 = 0
-
-	for ; data[pos] != '\r'; pos++ {
-		value = value*10 + int64(data[pos]-'0')
+// the intger value and the error
+// the function internally manipulates the buffer pointer
+// and keepts it at a point where the subsequent value can be read from.
+func readInt64(c io.ReadWriter, buf *bytes.Buffer) (int64, error) {
```

Comment:
Golang's Int is generally 32-bit wide on 32-bit machines and 64-bit wide on 64-bit machines. It is required to set the required width of Int.

```go
v, err := strconv.ParseInt(s,10,bits)
```
Providing an option to store 32-bit Int or Int32 on 64-bit targets could save memory as 4GB virtual memory space provided for a process could accommodate many values with no alignment issues.

Reply:
Though signed integer range would be affected

Reply:
RESP is for the client facing and hence we can go for 64-bit integers. While storing the data (in-mem and on-disk) we can allocate optimal space. This is precisely what Redis also does.

Reply:
I didn't know RESP had a defined integer width, just looked at it now. My bad.

Reply:
@rohanverma94 can you please share the doc where RESP has defined integer width, wasn't able to find any

Reply:
@asutosh97 This is from ``resp_parser.c``
![Screenshot from 2022-10-26 22-39-18](https://user-images.githubusercontent.com/8470490/198091907-5851402b-2c62-48dd-9c16-4c3c80951a00.png)

Though it says "parseLong()", but it converts to ``long long`` -->  at least 64 bits. 

Notice the function ```int string2ll(const char *s, size_t slen, long long *value)```

This works even if Redis is compiled against 32-bit as in ``make32``, this will still reduce the integer to 64-bit ``long long`` with a 32-bit code. It would still give allocators a headache but it gives some sort of standardization on integers. Though for perf it is recommended to use 32-bit data with 64-bit code. Salvatore made this decision citing that not many are going to compile a 32-bit version of Redis.

Reply:
thank you for the detailed context :)